batch/batch/driver/main.py

```diff
@@ -1322,6 +1330,139 @@ async def monitor_system(app):
+
+    next_key = await compact(None)
+    while next_key is not None:
+        await compact(next_key)
+
+
+async def compact_agg_billing_project_users_by_date_table(db: Database):
+    async def compact(key: Optional[Tuple[str, str, str, int]]) -> Optional[Tuple[str, str, str, int]]:
+        if key:
+            billing_date, billing_project, user, resource_id = key
+            where_condition = (
+                'AND ((billing_date > %s) OR'
+                '(billing_date = %s AND billing_project > %s) OR'
+                '(billing_date = %s AND billing_project = %s AND `user` > %s) OR'
+                '(billing_date = %s AND billing_project = %s AND `user` = %s AND resource_id > %s))'
+            )
+            args = [
+                billing_date,
+                billing_date,
+                billing_project,
+                billing_date,
+                billing_project,
+                user,
+                billing_date,
+                billing_project,
+                user,
+                resource_id,
+            ]
+        else:
+            where_condition = ''
+            args = []
+
+        token = secret_alnum_string(5)
+        scratch_table_name = f'`scratch_{token}`'
+
+        try:
+            next_key = await db.execute_and_fetchone(
+                f'''
+CREATE TEMPORARY TABLE {scratch_table_name} ENGINE=MEMORY
+AS (SELECT billing_date, billing_project, `user`, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+    FROM aggregated_billing_project_user_resources_by_date_v3
+    WHERE token != 0 {where_condition}
+    GROUP BY billing_date, billing_project, `user`, resource_id
+    ORDER BY billing_date, billing_project, `user`, resource_id
+    LIMIT 1000
+    LOCK IN SHARE MODE
+);
+
+DELETE aggregated_billing_project_user_resources_by_date_v3 FROM aggregated_billing_project_user_resources_by_date_v3
+INNER JOIN {scratch_table_name} ON aggregated_billing_project_user_resources_by_date_v3.billing_date = {scratch_table_name}.billing_date AND
+                      aggregated_billing_project_user_resources_by_date_v3.billing_project = {scratch_table_name}.billing_project AND
+                      aggregated_billing_project_user_resources_by_date_v3.`user` = {scratch_table_name}.`user` AND
+                      aggregated_billing_project_user_resources_by_date_v3.resource_id = {scratch_table_name}.resource_id
+WHERE token != 0;
+
+INSERT INTO aggregated_billing_project_user_resources_by_date_v3 (billing_date, billing_project, `user`, resource_id, token, `usage`)
+SELECT billing_date, billing_project, `user`, resource_id, 0, `usage`
+FROM {scratch_table_name}
+ON DUPLICATE KEY UPDATE `usage` = aggregated_billing_project_user_resources_by_date_v3.`usage` + {scratch_table_name}.`usage`;
+
+SELECT billing_date, billing_project, `user`, resource_id
+FROM {scratch_table_name}
+ORDER BY billing_date DESC, billing_project DESC, `user` DESC, resource_id DESC
+LIMIT 1;
+''',
```

Comment:
Could we possibly do this inside a transaction and assert that, for the locked rows, the total usage for a given key before and after the compaction is the same?

Reply:
I think an assertion of that sort would make me comfortable deploying this code, but it would be nice to follow up with some more observability. A couple of ideas that I think would be really illustrative:
- Sampling total usage for a billing project/user and see that usage is monotonically increasing over time
- Tracking a summary/histogram of how many rows/tokens exist for a particular key. On a long-running batch we should be able to see that rise and fall on a five-minute period.

Those two graphs together would give me a lot more confidence that this is working as expected, but I understand if you'd want to do that in parts.

Reply:
Ah, was this comment what lead to the sum-and-if? Maybe we should wait for Daniel to return and talk through his concerns.